### PR TITLE
[ALLUXIO-2818] Fix MasterFaultToleranceIntegrationTest 

### PR DESCRIPTION
**Pull request by peisun1115**

### CODE REVIEW

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 1]
```diff
@@ -226,11 +227,16 @@ public void start() throws IOException {
    */
   @Override
   public void stop() throws IOException {
+    // Use shutdownNow because HeartbeatThreads never finish until they are interrupted
+    getExecutorService().shutdownNow();
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: why is this reordered?

File: core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java [hunk 2]
```diff
@@ -174,6 +175,11 @@ public void stop() {
     if (mFilePersistenceService != null) {
       mFilePersistenceService.cancel(true);
     }
+    try {
+      getExecutorService().awaitTermination(10, TimeUnit.MINUTES);
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: Does the `getExecutorService().shutdownNow();` line need to be reordered like `DefaultBlockWorker`?

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java (hunk 1 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: I've added comment. What happened during the test is that: the heartbeat and just keep holding the lock (keep retry) due to the following code block which prevents FileSystemMaster.close from finishing.

`  public synchronized Set<Long> getPinList() throws ConnectionFailedException, IOException {
    return retryRPC(new RpcCallable<Set<Long>>() {
      @Override
      public Set<Long> call() throws TException {
        return mClient.getPinIdList();
      }
    });
  }`

File: core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java (hunk 2 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: done

File: core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java [hunk 3]
```diff
@@ -170,12 +171,22 @@ public void start() {
...
+    //    runnables running in the executors might be using the clients.
     mSessionCleaner.stop();
     if (mFilePersistenceService != null) {
       mFilePersistenceService.cancel(true);
     }
-    mFileSystemMasterWorkerClient.close();
     // This needs to be shutdownNow because heartbeat threads will only stop when interrupted.
     getExecutorService().shutdownNow();
+    try {
+      getExecutorService().awaitTermination(10, TimeUnit.MINUTES);
```
Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: Why 10 minutes? I would have expected something more like 5-10 seconds. Does this mean that it could take 10 minutes for a standby to upgrade to leader?

Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: check the return value

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMaster.java [hunk 4]
```diff
@@ -88,6 +89,9 @@ public void start() throws Exception {
           stopServing();
           stopMasters();
 
+          // TODO(peis): Stop serving properly and remove this master recreation.
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMaster.java]: why do we need to recreate masters?

File: core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java (hunk 3 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: I've changed this to a loop and add a comment.

Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: Similar as above, i've changed this to a loop and added a comment.

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMaster.java (hunk 4 above)
Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMaster.java]: I've removed this to stop/start more properly to avoid this master recreation. I had it initially because i thought this could fix the problem (it was recreating the masters before my journal change).

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java [hunk 5]
```diff
@@ -103,8 +103,13 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options, FileSystemContex
       mUnderStorageOutputStream = null;
     } else { // Write is through to the under storage, create mUnderStorageOutputStream
       try {
-        WorkerNetAddress worker = // not storing data to Alluxio, so block size is 0
-            options.getLocationPolicy().getWorkerForNextBlock(mBlockStore.getWorkerInfoList(), 0);
+        WorkerNetAddress worker;
+
+        while ((worker = // not storing data to Alluxio, so block size is 0
```
Comment (gpang) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: Why does this have to be in a loop, and could this loop forever? (not enough workers)

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java [hunk 6]
```diff
@@ -103,8 +103,13 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options, FileSystemContex
...
+            options.getLocationPolicy().getWorkerForNextBlock(mBlockStore.getWorkerInfoList(), 0))
+            == null) {
+        }
+        Preconditions.checkNotNull(worker, "No worker is found");
```
Comment (gpang) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: Is it always possible to have a worker to write to?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 7]
```diff
@@ -61,16 +62,16 @@
   private final int mMinWorkerThreads;
 
   /** The port for the RPC server. */
-  private final int mPort;
+  private int mPort;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: why is this no longer `final`?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 8]
```diff
@@ -118,15 +119,16 @@
...
-                  connectionTimeout);
+
+      mTServerSocket = new TServerSocket(NetworkAddressUtils.getBindAddress(ServiceType.MASTER_RPC),
+          Configuration.getInt(PropertyKey.MASTER_CONNECTION_TIMEOUT_MS));
       mPort = NetworkAddressUtils.getThriftPort(mTServerSocket);
       // reset master rpc port
       Configuration.set(PropertyKey.MASTER_RPC_PORT, Integer.toString(mPort));
       mRpcAddress = NetworkAddressUtils.getConnectAddress(ServiceType.MASTER_RPC);
 
+      mTransportProvider = TransportProvider.Factory.create();
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: Why is this moved down?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 9]
```diff
@@ -315,6 +330,8 @@ protected void startServingRPCServer() {
    * @throws Exception if the underlying jetty server throws an exception
    */
   protected void stopServing() throws Exception {
+    LOG.info("Alluxio master version {} stop serving @ {} (was serving: {})",
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: Are these messages duplicates?

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java (hunk 5 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: Oh, i need to remove this. This was a hack. I need to figure out a way to do this in the test.

**Comment by peisun1115** [comment]
Reviewers, I've marked this as WIP again since it has more problems.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 7 above)
Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: done.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 8 above)
Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: done

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 9 above)
Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: Removed.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 10]
```diff
@@ -208,6 +212,7 @@ protected void startMasters(boolean isLeader) {
       connectToUFS();
       mRegistry.start(isLeader);
     } catch (IOException e) {
+      LOG.error("Failed to start master.", e);
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: does the propagated exception not get logged? This looks like it would result in logging the same thing twice

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 11]
```diff
@@ -288,9 +293,18 @@ protected void startServingRPCServer() {
     try {
       transportFactory = mTransportProvider.getServerTransportFactory();
     } catch (IOException e) {
-      throw Throwables.propagate(e);
+      throw new RuntimeException(e);
     }
 
+    try {
+      if (mTServerSocket != null) {
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: should it be an error to start multiple times without stopping?

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 12]
```diff
@@ -227,11 +230,27 @@ public void start(WorkerNetAddress address) {
    */
   @Override
   public void stop() {
+    // Steps to shutdown:
+    // 1. Gracefully shutting down the runnables running in the executors.
```
Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: shutting -> shut

File: tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java [hunk 13]
```diff
@@ -56,11 +55,20 @@
 
   @After
   public final void after() throws Exception {
+    if (HadoopClientTestUtils.isHadoop1x()) {
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java]: I think we can replace all of the hadoop-1 checking with
```java
@BeforeClass
public static void beforeClass() {
  Assume.assumeFalse(HadoopClientTestUtils.isHadoop1x());
}
```

File: tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java [hunk 14]
```diff
@@ -229,20 +242,29 @@ public void killStandby() throws Exception {
...
     Assert.assertEquals(WORKER_CAPACITY_BYTES, store.getCapacityBytes());
 
     for (int kills = 0; kills < MASTERS - 1; kills++) {
       Assert.assertTrue(mMultiMasterLocalAlluxioCluster.stopLeader());
       mMultiMasterLocalAlluxioCluster.waitForNewMaster(CLUSTER_WAIT_TIMEOUT_MS);
-      waitForWorkerRegistration(store, 1, 5 * Constants.SECOND_MS);
-      // If worker is successfully re-registered, the capacity bytes should not change.
-      Assert.assertEquals(WORKER_CAPACITY_BYTES, store.getCapacityBytes());
+      long capacityFound = -1;
+      for (int i = 0; i < MASTERS; ++i) {
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java]: why do we loop over the masters?

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java [hunk 15]
```diff
@@ -238,13 +238,17 @@ public synchronized void flush() throws IOException {
...
+  public synchronized void close() {
+    try {
+      Closer closer = Closer.create();
+      if (mJournalOutputStream != null) {
+        closer.register(mJournalOutputStream);
+      }
+      closer.register(mGarbageCollector);
+      closer.close();
+    } catch (IOException e) {
+      LOG.warn("Failed to close UfsJournalLogWriter with error: {}.", e.getMessage());
```
Comment (calvinjia) [inline core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java]: What's the motivation for swallowing this exception, and is it really only a warning? Also, could we use `closeQuietly` (or an improved version if we want the logging)?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 16]
```diff
@@ -183,16 +184,19 @@ public Boolean apply(Void input) {
 
   @Override
   public void start() throws Exception {
+    LOG.info("Starting Alluxio master @{}", mRpcAddress);
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: Is this logging necessary, I think we are already handling it in `ProcessUtils` (though that might not work for unit tests)

File: core/server/worker/src/main/java/alluxio/worker/block/BlockMasterSync.java [hunk 17]
```diff
@@ -178,7 +178,7 @@ public void close() {
    * @throws Exception if an error occurs when executing the command
    */
   // TODO(calvin): Evaluate the necessity of each command.
-  private void handleMasterCommand(Command cmd) throws Exception {
+  private void handleMasterCommand(Command cmd) throws IOException, ConnectionFailedException {
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/BlockMasterSync.java]: Could you update the javadoc?

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 18]
```diff
@@ -227,11 +230,27 @@ public void start(WorkerNetAddress address) {
...
+    // 2. Shutdown the executors.
+    // 3. Shutdown the clients. This needs to happen after the executors is shutdown because
+    //    runnables running in the executors might be using the clients.
     mSessionCleaner.stop();
+    // The executor shutdown needs to be done in a loop with retry because the interrupt
+    // signal can sometimes be ignored.
+    CommonUtils.waitFor("shutdown block worker", new Function<Void, Boolean>() {
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: `block worker executor service shutdown`? (applies to other places as well)

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 10 above)
Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: done. Thanks.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 11 above)
Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: This is not null when the AlluxioMasterProcess is created since mTServerSocket is also created in the ctor (to get the socket address in case of 0-port).

File: tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java (hunk 13 above)
Comment (peisun1115) [inline tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java]: Thanks, this works great! I was trying Assume in 'before' and it didn't work.

File: tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java (hunk 14 above)
Comment (peisun1115) [inline tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java]: This was used in -Phadoop1 where a master can be transiently elected as master.  This was what happened:

1. Master A (leader) is stopped.
2. Master A is elected as leader.
3. Master A is shutdown because of the loop in LocalAlluxioMaster#stop. 

This only happened in hadoop1. I have not figured out why step 2 happened. Since we have disabled this test in hadoop1 profile, i am fine with removing it if you prefer.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 16 above)
Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: yeah, the integration test won't have this. maybe it is more scalable to add  this logging here. I've removed this and added corresponding logging in the local clusters.

File: core/server/worker/src/main/java/alluxio/worker/block/BlockMasterSync.java (hunk 17 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/block/BlockMasterSync.java]: done

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java (hunk 18 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: done

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java (hunk 15 above)
Comment (peisun1115) [inline core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java]: This was necessary for hadoop1. I've removed this. 

I think it was fine to swallow this because this is only called when a leader master is stopped. Leader master is usually stopped with an unexpected failure.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 19]
```diff
@@ -64,13 +65,13 @@
...
   /** The socket for thrift rpc server. */
-  private final TServerSocket mTServerSocket;
+  private TServerSocket mTServerSocket;
 
   /** The transport provider to create thrift server transport. */
   private final TransportProvider mTransportProvider;
 
   /** The address for the rpc server. */
-  private final InetSocketAddress mRpcAddress;
+  private InetSocketAddress mRpcAddress;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: NIT: Is this allowed to stay `final`?

Comment (peisun1115) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: done. yes. thanks.

**Comment by peisun1115** [comment]
@gpang PTAL? thanks

File: tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java (hunk 14 above)
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java]: I'd prefer to remove it, but I'd also be fine with leaving it and adding a detailed comment

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 11 above)
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: got it

File: tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java (hunk 14 above)
Comment (peisun1115) [inline tests/src/test/java/alluxio/master/MasterFaultToleranceIntegrationTest.java]: k. removed
